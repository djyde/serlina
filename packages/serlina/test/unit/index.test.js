const {
  Serlina
} = require('../../dist')
const path = require('path')

const GLOBAL_PAYLOAD = {
  foo: 'foo'
}
const PAGE1 = 'page1' // don't have style
const PAGE2 = 'page2' // have style

const PUBLIC_PATH = '/public/'
const BASE_DIR = path.resolve(__dirname, './fixtures/sample')

describe('Serlina', () => {

  describe('dev mode', () => {

    let app

    beforeAll(async () => {
      app = new Serlina({
        baseDir: BASE_DIR,
        __testing: true
      })
      await app.prepare()
      app.inject(GLOBAL_PAYLOAD)
    })

    test('snapshot', async () => {
      const rendered = await app.render(PAGE1)
      expect(rendered.body).toMatchSnapshot()
      const rendered2 = await app.render(PAGE2)
      expect(rendered2.body).toMatchSnapshot()
    })

    test('inject', async () => {
      const payload = {
        bar: 'bar'
      }
      const rendered = await app.render(PAGE1, payload)
      expect(rendered.__injected).toEqual(Object.assign({}, GLOBAL_PAYLOAD, payload))
    })

    test('page assets', async () => {
      const rendered = await app.render(PAGE1)
      expect(rendered.__pageScripts).toEqual([
        PAGE1 + '.js',
        '_SERLINA_VENDOR.js',
        '_SERLINA_MAIN.js'
      ])

      expect(rendered.__pageStyles).toEqual([])
    })

    test('page assets (have style)', async () => {
      const rendered = await app.render(PAGE2)
      expect(rendered.__pageScripts).toEqual([
        PAGE2 + '.js',
        '_SERLINA_VENDOR.js',
        '_SERLINA_MAIN.js'
      ])

      expect(rendered.__pageStyles).toEqual([
        PAGE2 + '.css'
      ])
    })

    test('get initial props', async () => {
      const rendered = await app.render(PAGE2)
      expect(rendered.__initialProps).toEqual({
        foo: 'foo'
      })
    })

    test('render a non exist page', async () => {
      const rendered = await app.render('non-exist')

      expect(rendered.__pageScripts).toEqual([
        '_404.js',
        '_SERLINA_VENDOR.js',
        '_SERLINA_MAIN.js'
      ])
    })

    test('nested page', async () => {
      const rendered = await app.render('user/list')
      expect(rendered.body).toMatchSnapshot()
    })
  })

  describe('prod mode', () => {

    let app;
    let assetsMap;

    beforeAll(() => {
      app = new Serlina({
        baseDir: BASE_DIR,
        dev: false,
        publicPath: PUBLIC_PATH,
        __testing: true
      })
      app.inject(GLOBAL_PAYLOAD)
      return new Promise((res, rej) => {
        const compiler = app.build()
        compiler.run((err, stats) => {
          if (err || stats.hasErrors()) {
            rej(err || stats.toJson())
          } else {
            assetsMap = require(path.resolve(BASE_DIR, './.serlina/assetsmap.json'))
            app.prepare().then(res(stats.toJson())).catch(rej)
          }
        })
      })
    }, 20000)

    test('page assets', async () => {
      const rendered = await app.render(PAGE1)
      expect(rendered.__pageScripts).toEqual([
        assetsMap[PAGE1].js,
        assetsMap['_SERLINA_VENDOR'].js,
        assetsMap['_SERLINA_MAIN'].js
      ])

      expect(rendered.__pageStyles).toEqual([])
    })

    test('page assets (have style)', async () => {
      const rendered = await app.render(PAGE2)
      expect(rendered.__pageScripts).toEqual([
        assetsMap[PAGE2].js,
        assetsMap['_SERLINA_VENDOR'].js,
        assetsMap['_SERLINA_MAIN'].js
      ])

      expect(rendered.__pageStyles).toEqual([
        assetsMap[PAGE2].css
      ])
    })

    test('render a non exist page', async () => {
      const rendered = await app.render('non-exist')

      expect(rendered.__pageScripts).toEqual([
        assetsMap['_SERLINA_VENDOR'].js,
        assetsMap['_SERLINA_MAIN'].js
      ])
    })

    test('nested page', async () => {
      const rendered = await app.render('user/list')
      expect(rendered.body)
    })

  })

  describe('test config', () => {

    test('webpack config', async () => {

      const app = new Serlina({
        baseDir: BASE_DIR,
        __testing: true
      })
      await app.prepare()
      app.inject(GLOBAL_PAYLOAD)

      const [serverSide, clientSide, vendors] = app._webpackConfig

      // TODO: webpack config test
      expect(serverSide)
      expect(clientSide)
      expect(vendors)
    })

  })

  describe('inline css', () => {
    test('inline css', async () => {
      const app = new Serlina({
        baseDir: BASE_DIR,
        __testing: true,
        inlineCSS: true,
      })
      await app.prepare()
      app.inject(GLOBAL_PAYLOAD)

      const rendered = await app.render('page2')
      expect(rendered.body).toMatchSnapshot()
    })
  })
})