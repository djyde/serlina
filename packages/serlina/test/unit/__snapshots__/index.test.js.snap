// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serlina dev mode nested page 1`] = `
"<!DOCTYPE html><html><head><title data-react-helmet=\\"true\\"></title><script>
        window.__serlina__DATA = {};
        window.__serlina__DATA.pageInitialProps = {};
      </script></head><body><div id=\\"app\\"><div data-reactroot=\\"\\">User List</div></div><script src=\\"http://127.0.0.1:3000/_SERLINA_VENDOR.js\\"></script><script src=\\"http://127.0.0.1:3000/user/list.js\\"></script><script src=\\"http://127.0.0.1:3000/_SERLINA_MAIN.js\\"></script></body></html>"
`;

exports[`Serlina dev mode snapshot 1`] = `
"<!DOCTYPE html><html><head><title data-react-helmet=\\"true\\"></title><script>
        window.__serlina__DATA = {};
        window.__serlina__DATA.pageInitialProps = {};
      </script></head><body><div id=\\"app\\"><div data-reactroot=\\"\\">Page1</div></div><script src=\\"http://127.0.0.1:3000/_SERLINA_VENDOR.js\\"></script><script src=\\"http://127.0.0.1:3000/page1.js\\"></script><script src=\\"http://127.0.0.1:3000/_SERLINA_MAIN.js\\"></script></body></html>"
`;

exports[`Serlina dev mode snapshot 2`] = `
"<!DOCTYPE html><html><head><title data-react-helmet=\\"true\\"></title><link rel=\\"stylesheet\\" href=\\"http://127.0.0.1:3000/page2.css\\"/><script>
        window.__serlina__DATA = {};
        window.__serlina__DATA.pageInitialProps = {\\"foo\\":\\"foo\\"};
      </script></head><body><div id=\\"app\\"><div data-reactroot=\\"\\">Page2 - <!-- -->foo</div></div><script src=\\"http://127.0.0.1:3000/_SERLINA_VENDOR.js\\"></script><script src=\\"http://127.0.0.1:3000/page2.js\\"></script><script src=\\"http://127.0.0.1:3000/_SERLINA_MAIN.js\\"></script></body></html>"
`;

exports[`Serlina inline css inline css 1`] = `
"<!DOCTYPE html><html><head><title data-react-helmet=\\"true\\"></title><style>body {
  background: red;
}
</style><script>
        window.__serlina__DATA = {};
        window.__serlina__DATA.pageInitialProps = {\\"foo\\":\\"foo\\"};
      </script></head><body><div id=\\"app\\"><div data-reactroot=\\"\\">Page2 - <!-- -->foo</div></div><script src=\\"http://127.0.0.1:3000/_SERLINA_VENDOR.js\\"></script><script src=\\"http://127.0.0.1:3000/page2.js\\"></script><script src=\\"http://127.0.0.1:3000/_SERLINA_MAIN.js\\"></script></body></html>"
`;
